### ISSUE TITLE
[spec] i18n revamp (WORK IN PROGRESS) 

### ISSUE THREAD

**Issue by amcgee** [dhis2/app-platform#369]
**THIS IS A WORK IN PROGRESS**, I will continue fleshing it out before sharing for comment

I'm opening this issue to discuss an overhaul of the internationalization (**i18n**) system.  I'll start by laying out the challenges we face with the current system, and this issue will then coalesce into a specification for an overhaul of the platform-native i18n functionality.

#  The Problem(s)

## Types of translations in DHIS2

There are three primary categories of strings in 
1. **Server-bundled translations** For a long time, all translations were bundled into the DHIS2 core and accessible through an API endpoint.  Pretty much all use of this system has been eliminated, but some strings still have their translations baked into the WAR file (notably bundled application names)

2. **Application-bundled translations** 

3. **Dynamic translations** Some strings are truly dynamic, created by the implementer of the instance when configuring their metadata model.  These strings are created, updated, and accessed through the API and stored in the database.  We won't be directly addressing this use-case in for application-bundled strings, but it's good to keep in mind.

## A brief history of application-bundled translations

1. **d2 (the ancient solution)** -

2. **d2-i18n (the legacy solution)** - 

3. **d2-app-scripts i18n (the current solution)** - 

There are a few straggling uses of the (relatively) ancient `d2` system in some of our libraries (notably `d2-ui`).  I won't address the challenges with 

## Challenges with the `d2-i18n` "legacy" solution

Until very recently, the "standard" fleet of internationalization tools for DHIS2 libraries and applications was :

- [d2-i18n](https://github.com/dhis2/d2-i18n)
- [d2-i18n-extract](https://github.com/dhis2/d2-i18n-extract)
- [d2-i18n-generate](https://github.com/dhis2/d2-i18n-generate)

...TODO

## Challenges with the "current" `d2-app-scripts i18n` system

This is almost a direct port of the previous system.  It combines ported versions of `d2-i18n-extract` and `d2-i18n-generate` into a single App Platform cli module (`d2-app-scripts i18n`) and automatically runs the extract and generate steps when building an application or library with the platform.  The components are:

- [d2-i18n](https://github.com/dhis2/d2-i18n)
- [d2-app-scripts i18n extract](https://github.com/dhis2/app-platform/tree/master/cli/src/lib/i18n/extract.js)
- [d2-app-scripts i18n generate](https://github.com/dhis2/app-platform/tree/master/cli/src/lib/i18n/generate.js)

...TODO

# [SPEC] Requirements for a revamped i18n solution

...TODO

# A proposed design

...TODO

# Things left unaddressed by this spec

...TODO

**Comment by ismay** [comment]
We discussed some i18n things @Mohammer5 and I encountered [here on slack](https://dhis2.slack.com/archives/CLJSVA1RV/p1588080988221700). We were discussing the usecase for namespaces, and @amcgee noted that even though the string might be the same, the context could mean that the translation would be different. But we don't really provide context to translators. So one thing we could add is that we should start adding context for translators.

**Comment by amcgee** [comment]
Related - static analysis of code to find untranslated UI strings https://github.com/dhis2/cli-style/issues/256 (Thanks @ismay @martinkrulltott @HendrikThePendric)

**Comment by amcgee** [comment]
See also https://github.com/dhis2/app-platform/pull/98 (closed)

> I went halfway down the road of building a babel plugin to inject the namespace into the string key of i18n.t calls, but I stopped because it might not be the right approach - we may instead want to stub out the import ‘@dhis2/d2-i18n’ module and replace it with something we control, something that has a fixedT with a set namespace. Here’s my half-complete plugin, worth further investigation...

https://astexplorer.net/#/gist/a5bd7b8c733fd52e3b51d3f713b5a3d3/6b34483cb68fec86c57bd6c3cf7b32dd4995871e (edited)

**Comment by Mohammer5** [comment]
Another requirement proposal:

It'd be nice to be able to use markup / react components inside the translation string. i18n-react (or so) has a solution for this already. @cooper-joe made some designs for the new maintenance app where there were some styled withing translatable text (like bold text)

**Comment by amcgee** [comment]
> It'd be nice to be able to use markup / react components inside the translation string. i18n-react (or so) has a solution for this already. @cooper-joe made some designs for the new maintenance app where there were some styled withing translatable text (like bold text)

@Mohammer5 definitely!  The challenge with `react-i18next` was at one point that it didn't have good support for right-to-left text, I believe, but it's worth re-evaluating!

**Comment by ismay** [comment]
See also this thread: https://dhis2.slack.com/archives/C0BP0RABF/p1601970390067700. We also need to support localised pluralization, and allowing for translation context would be good as well. Currently I don't think we properly support that (but correct me if I'm wrong).

**Comment by jenniferarnesen** [comment]
Regarding plurals, there is a problem with our current set up that is short-circuiting translations. This is currently happening for the interpretations component in d2-ui.

Locally, the `en.pot` file is extracted from the code. Since English uses the keys as the translation, and plural keys have a "_plural" appended like this: `{{count}} reply_plural`, our code needs to actually provide English translations for the singular and plural versions of the string. This can be done in the code like this:  

```
const replies = i18n.t("{{count}} reply", {
        count: repliedBy.length,
        defaultValue: "{{count}} reply",
        defaultValue_plural: "{{count}} replies" 
    });
```

Once extracted, the en.pot file looks like this:
```
msgid "{{count}} reply"
msgid_plural "{{count}} reply"
msgstr[0] "{{count}} reply"
msgstr[1] "{{count}} replies"
```

But transifex does not accept pot files with translations in them (the msgstr[0] and msgstr[1]). So currently, the process stops here and updated po files are not being created on transifex.
